### DISCUSSION TITLE
linking.release 报错

### DISCUSSION THREAD

**Discussion by Mybuhe** [xmake-io/xmake#3356] [Q&A]
这是xmake.lua 文件

add_rules("mode.debug", "mode.release")

add_requires("abseil 20220623.0")
add_requires("libev 4.33")
add_requires("yaml-cpp 0.7.0")
add_requires("openssl 1.1.1-t", {alias = "openssl", configs = {options = "OpenSSL:shared=True"}})
set_languages("gnuxx14")

add_includedirs("sub/rtcbase/src","src")

target("rtcbase")
    set_kind("static")
    set_policy("build.across_targets_in_parallel", false)
    add_defines("WEBRTC_POSIX","WEBRTC_LINUX","NDEBUG")
    add_files(
        "sub/rtcbase/src/rtc_base/*.cc",
        "sub/rtcbase/src/rtc_base/deprecated/*.cc",
        "sub/rtcbase/src/rtc_base/internal/*.cc",
        "sub/rtcbase/src/rtc_base/strings/*.cc",
        "sub/rtcbase/src/rtc_base/synchronization/*.cc",
        "sub/rtcbase/src/rtc_base/task_utils/*.cc",
        "sub/rtcbase/src/rtc_base/third_party/base64/*.cc",
        "sub/rtcbase/src/rtc_base/third_party/sigslot/*.cc",
        "sub/rtcbase/src/api/crypto/*.cc",
        "sub/rtcbase/src/api/task_queue/*.cc",
        "sub/rtcbase/src/api/units/*.cc",
        "sub/rtcbase/src/system_wrappers/source/*.cc"
    )
    add_packages("abseil", "openssl")

target("untitled")
    set_kind("binary")
    add_syslinks("pthread")
    add_files("src/*.cpp","src/base/*.cpp","src/server/*.cpp")
    add_deps("rtcbase") 
    add_packages("libev", "yaml-cpp", "abseil")

报错是

build/linux/x86_64/release/librtcbase.a(string_encode.cc.o): In function `rtc::tokenize_first(absl::lts_20220623::string_view, char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
string_encode.cc:(.text+0x6bb): undefined reference to `absl::lts_20220623::string_view::find(char, unsigned long) const'
collect2: error: ld returned 1 exit status

构建日志是
 linking.release untitled
/usr/bin/g++ -o build/linux/x86_64/release/untitled build/.objs/untitled/linux/x86_64/release/src/main.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/conf.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/event_loop.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/log.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/socket.cpp.o build/.objs/untitled/linux/x86_64/release/src/server/signaling_server.cpp.o -m64 -L/home/buhe/.xmake/packages/l/libev/4.33/0d0a942dbe0549e892568b8afdf51ad9/lib -L/home/buhe/.xmake/packages/y/yaml-cpp/0.7.0/c43d7f3fd9044adcaea2b25c200c62dd/lib -Lbuild/linux/x86_64/release -L/home/buhe/.xmake/packages/a/abseil/20220623.0/c60e4aa2b7614ef29c9a1e17127fd352/lib -L/home/buhe/.xmake/packages/o/openssl/1.1.1-t/25a06d63c5c94540b3812b3f969192cb/lib -s -lev -lyaml-cpp -lrtcbase -labsl_status -labsl_cord -labsl_flags -labsl_flags_parse -labsl_flags_internal -labsl_flags_reflection -labsl_flags_marshalling -labsl_flags_commandlineflag_internal -labsl_synchronization -labsl_time -labsl_hash -labsl_city -labsl_time_zone -labsl_spinlock_wait -labsl_failure_signal_handler -labsl_bad_optional_access -labsl_flags_commandlineflag -labsl_random_internal_pool_urbg -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_cordz_handle -labsl_cordz_sample_token -labsl_base -labsl_bad_any_cast_impl -labsl_periodic_sampler -labsl_random_distributions -labsl_flags_usage_internal -labsl_random_seed_sequences -labsl_throw_delegate -labsl_stacktrace -labsl_symbolize -labsl_debugging_internal -labsl_flags_private_handle_accessor -labsl_strings -labsl_flags_config -labsl_malloc_internal -labsl_str_format_internal -labsl_flags_usage -labsl_strings_internal -labsl_flags_program_name -labsl_int128 -labsl_scoped_set_env -labsl_raw_hash_set -labsl_random_internal_seed_material -labsl_random_internal_randen -labsl_random_internal_randen_slow -labsl_random_internal_randen_hwaes_impl -labsl_random_internal_randen_hwaes -labsl_graphcycles_internal -labsl_exponential_biased -labsl_bad_variant_access -labsl_statusor -labsl_random_internal_distribution_test_util -labsl_random_internal_platform -labsl_hashtablez_sampler -labsl_demangle_internal -labsl_leak_check -labsl_log_severity -labsl_raw_logging_internal -labsl_strerror -labsl_examine_stack -labsl_low_level_hash -labsl_random_seed_gen_exception -labsl_civil_time -lssl -lcrypto -lpthread -ldl
build/linux/x86_64/release/librtcbase.a(string_encode.cc.o): In function `rtc::tokenize_first(absl::lts_20220623::string_view, char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
string_encode.cc:(.text+0x6bb): undefined reference to `absl::lts_20220623::string_view::find(char, unsigned long) const'
collect2: error: ld returned 1 exit status
error: @programdir/modules/private/async/runjobs.lua:256: @programdir/actions/build/kinds/binary.lua:74: @programdir/core/sandbox/modules/os.lua:373: execv(/usr/bin/g++ -o build/linux/x86_64/release/untitled build/.objs/untitled/linux/x86_64/release/src/main.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/conf.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/event_loop.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/log.cpp.o build/.objs/untitled/linux/x86_64/release/src/base/socket.cpp.o build/.objs/untitled/linux/x86_64/release/src/server/signaling_server.cpp.o -m64 -L/home/buhe/.xmake/packages/l/libev/4.33/0d0a942dbe0549e892568b8afdf51ad9/lib -L/home/buhe/.xmake/packages/y/yaml-cpp/0.7.0/c43d7f3fd9044adcaea2b25c200c62dd/lib -Lbuild/linux/x86_64/release -L/home/buhe/.xmake/packages/a/abseil/20220623.0/c60e4aa2b7614ef29c9a1e17127fd352/lib -L/home/buhe/.xmake/packages/o/openssl/1.1.1-t/25a06d63c5c94540b3812b3f969192cb/lib -s -lev -lyaml-cpp -lrtcbase -labsl_status -labsl_cord -labsl_flags -labsl_flags_parse -labsl_flags_internal -labsl_flags_reflection -labsl_flags_marshalling -labsl_flags_commandlineflag_internal -labsl_synchronization -labsl_time -labsl_hash -labsl_city -labsl_time_zone -labsl_spinlock_wait -labsl_failure_signal_handler -labsl_bad_optional_access -labsl_flags_commandlineflag -labsl_random_internal_pool_urbg -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_cordz_handle -labsl_cordz_sample_token -labsl_base -labsl_bad_any_cast_impl -labsl_periodic_sampler -labsl_random_distributions -labsl_flags_usage_internal -labsl_random_seed_sequences -labsl_throw_delegate -labsl_stacktrace -labsl_symbolize -labsl_debugging_internal -labsl_flags_private_handle_accessor -labsl_strings -labsl_flags_config -labsl_malloc_internal -labsl_str_format_internal -labsl_flags_usage -labsl_strings_internal -labsl_flags_program_name -labsl_int128 -labsl_scoped_set_env -labsl_raw_hash_set -labsl_random_internal_seed_material -labsl_random_internal_randen -labsl_random_internal_randen_slow -labsl_random_internal_randen_hwaes_impl -labsl_random_internal_randen_hwaes -labsl_graphcycles_internal -labsl_exponential_biased -labsl_bad_variant_access -labsl_statusor -labsl_random_internal_distribution_test_util -labsl_random_internal_platform -labsl_hashtablez_sampler -labsl_demangle_internal -labsl_leak_check -labsl_log_severity -labsl_raw_logging_internal -labsl_strerror -labsl_examine_stack -labsl_low_level_hash -labsl_random_seed_gen_exception -labsl_civil_time -lssl -lcrypto -lpthread -ldl) failed(1)
stack traceback:
    [C]: in function 'error'
    [@programdir/core/base/os.lua:898]:
    [@programdir/core/sandbox/modules/os.lua:373]: in function 'execv'
    [@programdir/modules/core/tools/gcc.lua:398]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]:
    [@programdir/actions/build/kinds/binary.lua:74]: in function 'callback'
    [@programdir/modules/core/project/depend.lua:193]: in function 'on_changed'
    [@programdir/actions/build/kinds/binary.lua:55]: in function '_do_link_target'
    [@programdir/actions/build/kinds/binary.lua:102]:
    [@programdir/actions/build/kinds/binary.lua:129]: in function '_link_target'
    [@programdir/actions/build/kinds/binary.lua:157]: in function 'jobfunc'
    [@programdir/modules/private/async/runjobs.lua:232]:
    [C]: in function 'xpcall'
    [@programdir/core/base/utils.lua:280]: in function 'trycall'
    [@programdir/core/sandbox/modules/try.lua:117]: in function 'try'
    [@programdir/modules/private/async/runjobs.lua:218]: in function 'cotask'
    [@programdir/core/base/scheduler.lua:404]:

stack traceback:
        [C]: in function 'error'
        @programdir/core/base/os.lua:898: in function 'os.raiselevel'
        (...tail calls...)
        @programdir/modules/private/async/runjobs.lua:256: in field 'catch'
        @programdir/core/sandbox/modules/try.lua:123: in global 'try'
        @programdir/modules/private/async/runjobs.lua:218: in upvalue 'cotask'
        @programdir/core/base/scheduler.lua:404: in function <@programdir/core/base/scheduler.lua:397

**Comment by Mybuhe** [comment]
拜托大神，拜托大神

**Reply by waruqi** [reply]
估计是 abseil 包的子库 links 顺序问题，你可以到 xmake-repo/abseil 包里 用 add_links 调整下链接顺序试试

回头有时间我再看下

**Reply by Mybuhe** [reply]
那我按照cmake 去排一下顺序

**Comment by waruqi** [comment]
这个库其实有排序过的，但是里面的部分子库的顺序可能还不完全对 

你可以根据你现在遇到的符号错误，分析下部分子库的依赖关系，再调整下，提个 pr 过来

改下这里

https://github.com/xmake-io/xmake-repo/blob/944b3d6f0491d0d5f40cebdd779f16aa3630b575/packages/a/abseil/xmake.lua#L17

**Reply by Mybuhe** [reply]
我单独把absl_strings库给拿出来，把他下面的子库都排序了一下，还是不行.....，应该是我项目的问题........

**Reply by Mybuhe** [reply]
我是这么解决的   set_languages("gnuxx14","gnuxx17","cxx17")

**Reply by Mybuhe** [reply]
gcc 必须要11吧...才支持c17？

**Reply by Mybuhe** [reply]
我的项目本身的问题
